recipes/enable_chef_error_handler.rb

```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+#
+# Cookbook:: aws-parallelcluster
+# Recipe:: enable_chef_error_handler
+#
+# Copyright:: 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+chef_handler 'WriteChefError::WriteChefError' do
+  source "/etc/chef/cookbooks/aws-parallelcluster/files/default/event_handler/write_chef_error_handler.rb"
+  type exception: true
```

Comment:
According to [official doc](https://docs.chef.io/resources/chef_handler/#:~:text=a%20Ruby%20gem.-,type,of%20handler%20to%20register%20as%2C%20i.e.%20%3Areport%2C%20%3Aexception%20or%20both.,-Common%20Resource%20Functionality) by default both exception and report are enabled.

Probably you have to use `type :report => false, :exception => true` and after doing this you can remove the `if run_status.failed?` check in the handler, because it will be called only at exception case.

Reply:
Yes, according to the doc, there are 3 types of handlers: exception, report, and start. Of them,
- An exception handler runs when the `failed?` property for the `run_status` object returns `true`.
- A report handler runs when the `success?` property for the `run_status` object returns `true`.

Do you know what's the difference between 
(1) specifying `type :report => false, :exception => true` in `enable_chef_error_handler.rb`, and 
(2) change `def report` to `def exception` in `write_chef_error_handler.rb`?

I can play around with it (maybe next week), but wonder: if there're different ways to configure what we want, how do we know which is the best way?

Reply:
I paused on this activity as I was back on the design doc last month. I'm now back on the Chef handler task and trying to address this comment within the new PR here: https://github.com/aws/aws-parallelcluster-cookbook/pull/1641

1. My conclusion 1: There is no need to specify `{type :report => false, :exception => true}`. It is true that the default value for the "type" property is `{"report"=>true, "exception"=>true}`, when it's _not specified_. As long as I specify it (for example here, using `type exception: true`), it will replace the default value _as a whole_ (note: NOT just partially replace the `"exception"=>true` and leave `"report"=>true`). In this case here, by specifying `type exception: true`, my handler will only be registered as an exception handler.
2. Support for conclusion 1: I checked chef-client logs for failed runs and successful runs, and confirmed that this handler will only be executed when the run fails:

- When the run fails, the handler `WriteChefError::WriteChefError` will be executed:
```
Running handlers:
[2023-01-13T01:36:58+00:00] ERROR: Running exception handlers
[2023-01-13T01:36:58+00:00] INFO: run_status.formatted_exception:
[2023-01-13T01:36:58+00:00] INFO: 'Timeout::Error: execution expired'
......
  - WriteChefError::WriteChefError
Running handlers complete
[2023-01-13T01:36:58+00:00] ERROR: Exception handlers complete
```
- When the run succeeds, the handler `WriteChefError::WriteChefError` will not be executed:
```
Running handlers:
[2023-01-12T04:08:20+00:00] INFO: Running report handlers
Running handlers complete
[2023-01-12T04:08:20+00:00] INFO: Report handlers complete
```
3. This stays true even after I removed the `if run_status.failed?` check. So this comes to my 2nd conclusion: the `if run_status.failed?` check can be safely removed since the handler is an exception handler and will only be executed when the run fails.
4. My previous thought of changing `def report` to `def exception` would probably not work, because the type of the handler (whether a report handler or exception handler) is not defined by the name of the function.
5. Based on this I will remove the `if run_status.failed?` check in the new PR (https://github.com/aws/aws-parallelcluster-cookbook/pull/1641).

Does it address your comment @enrico-usai?